cores/esp8266/core_esp8266_i2s.c

```diff
@@ -30,147 +30,278 @@
+printf("%d: %d %d %d %d %d %d %p %p\n", i, rx->slc_items[i].myid, rx->slc_items[i].owner, rx->slc_items[i].eof, rx->slc_items[i].sub_sof, rx->slc_items[i].datalen, rx->slc_items[i].blocksize,
+rx->slc_items[i].buf_ptr, rx->slc_items[i].next_link_ptr);
+}
+}
+#endif
+
+
+static void ICACHE_FLASH_ATTR i2s_slc_begin() {
+  if (tx) {
+    _alloc_channel(tx);
+  }
+  if (rx) {
+    _alloc_channel(rx);
   }
 
   ETS_SLC_INTR_DISABLE();
   SLCC0 |= SLCRXLR | SLCTXLR;
   SLCC0 &= ~(SLCRXLR | SLCTXLR);
   SLCIC = 0xFFFFFFFF;
 
-  //Configure DMA
-  SLCC0 &= ~(SLCMM << SLCM); //clear DMA MODE
-  SLCC0 |= (1 << SLCM); //set DMA MODE to 1
-  SLCRXDC |= SLCBINR | SLCBTNR; //enable INFOR_NO_REPLACE and TOKEN_NO_REPLACE
-  SLCRXDC &= ~(SLCBRXFE | SLCBRXEM | SLCBRXFM); //disable RX_FILL, RX_EOF_MODE and RX_FILL_MODE
+  // Configure DMA
+  SLCC0 &= ~(SLCMM << SLCM); // Clear DMA MODE
+  SLCC0 |= (1 << SLCM); // Set DMA MODE to 1
+  SLCRXDC |= SLCBINR | SLCBTNR; // Enable INFOR_NO_REPLACE and TOKEN_NO_REPLACE
+  SLCRXDC &= ~(/*SLCBRXFE |*/ SLCBRXEM | SLCBRXFM); //disable RX_FILL, RX_EOF_MODE and RX_FILL_MODE
 
   //Feed DMA the 1st buffer desc addr
   //To send data to the I2S subsystem, counter-intuitively we use the RXLINK part, not the TXLINK as you might
   //expect. The TXLINK part still needs a valid DMA descriptor, even if it's unused: the DMA engine will throw
   //an error at us otherwise. Just feed it any random descriptor.
   SLCTXL &= ~(SLCTXLAM << SLCTXLA); // clear TX descriptor address
-  SLCTXL |= (uint32)&i2s_slc_items[1] << SLCTXLA; //set TX descriptor address. any random desc is OK, we don't use TX but it needs to be valid
   SLCRXL &= ~(SLCRXLAM << SLCRXLA); // clear RX descriptor address
-  SLCRXL |= (uint32)&i2s_slc_items[0] << SLCRXLA; //set RX descriptor address
+  if (!rx) {
+    SLCTXL |= (uint32)&tx->slc_items[1] << SLCTXLA; // Set fake (unused) RX descriptor address
+  } else {
+    SLCTXL |= (uint32)&rx->slc_items[0] << SLCTXLA; // Set real RX address
+  }
+  if (!tx) {
+//    SLCRXL |= (uint32)&rx->slc_items[1] << SLCRXLA; // Set fake (ununsed) TX descriptor address
+  } else {
+    SLCRXL |= (uint32)&tx->slc_items[0] << SLCRXLA; // Set real TX address
+  }
 
   ETS_SLC_INTR_ATTACH(i2s_slc_isr, NULL);
-  SLCIE = SLCIRXEOF; //Enable only for RX EOF interrupt
+  SLCIE = (tx?SLCIRXEOF:0) | (rx?SLCITXEOF:0); // Enable appropriate EOF IRQ
 
   ETS_SLC_INTR_ENABLE();
 
-  //Start transmission
+  // Start transmission ("TX" DMA always needed to be enabled)
   SLCTXL |= SLCTXLS;
-  SLCRXL |= SLCRXLS;
+  if (tx) {
+    SLCRXL |= SLCRXLS;
+  }
 }
 
-void ICACHE_FLASH_ATTR i2s_slc_end(){
+static void ICACHE_FLASH_ATTR i2s_slc_end(){
   ETS_SLC_INTR_DISABLE();
   SLCIC = 0xFFFFFFFF;
   SLCIE = 0;
   SLCTXL &= ~(SLCTXLAM << SLCTXLA); // clear TX descriptor address
   SLCRXL &= ~(SLCRXLAM << SLCRXLA); // clear RX descriptor address
 
   for (int x = 0; x<SLC_BUF_CNT; x++) {
-	  free(i2s_slc_buf_pntr[x]);
+    if (tx) {
```

Comment:
Not really relevant in this particular case, but:
Rather than a for cycle with two if statements inside, it's better to have two if statements with each a for cycle inside.